src/guardian/keys-validation/keys-validation.service.ts

```diff
@@ -46,91 +43,92 @@ export class KeysValidationService {
     );
     const genesisForkVersion: Uint8Array = await this.forkVersion();
     const genesisForkVersionBuffer = Buffer.from(genesisForkVersion.buffer);
-    const depositDataList = this.createDepositDataList(
-      keys,
-      withdrawalCredentialsBuffer,
-      genesisForkVersionBuffer,
-    );
-    return await this.findInvalidKeys(keys, depositDataList);
-  }
 
-  async findInvalidKeys(
-    keys: RegistryKey[],
-    depositDataList: DepositData[],
-  ): Promise<RegistryKey[]> {
-    const validatedKeys = await this.validateKeys(depositDataList);
+    const { cachedInvalidKeyList, uncachedDepositKeyList } =
+      this.partitionCachedData(
+        keys,
+        withdrawalCredentialsBuffer,
+        genesisForkVersionBuffer,
+      );
+
+    this.logger.log('Validation status of deposit keys:', {
+      cachedInvalidKeyCount: cachedInvalidKeyList.length,
+      keysNeedingValidationCount: uncachedDepositKeyList.length,
+      totalKeysCount: keys.length,
+    });
+
+    const validatedDepositKeyList: [DepositKey & Key, boolean][] =
+      await this.keyValidator.validateKeys<DepositKey>(uncachedDepositKeyList);
 
+    this.updateCache(validatedDepositKeyList);
+
+    const invalidKeys = this.filterInvalidKeys(validatedDepositKeyList);
+
+    return [...cachedInvalidKeyList, ...invalidKeys];
```

Comment:
Better to use `concat` here to join arrays, because `...` can cause stack overflow on big arrays.